lib/pure/bitops.nim

```diff
@@ -727,89 +730,259 @@ proc countTrailingZeroBits*(x: SomeInteger): int {.inline, noSideEffect.} =
+    when defined(amd64):
+      func builtin_rotl64(value: culonglong, shift: cint): culonglong
+                         {.importc: "__rolq", header: "immintrin.h".}
+
+    func builtin_rotr8(value: cuchar, shift: cint): cuchar
+                      {.importc: "__rorb", header: "immintrin.h".}
+    func builtin_rotr16(value: cushort, shift: cint): cushort
+                       {.importc: "__rorw", header: "immintrin.h".}
+    func builtin_rotr32(value: cuint, shift: cint): cuint
+                       {.importc: "__rord", header: "immintrin.h".}
+    when defined(amd64):
+      func builtin_rotr64(value: culonglong, shift: cint): culonglong
+                         {.importc: "__rorq", header: "immintrin.h".}
+
+# https://blog.regehr.org/archives/1063
+# https://stackoverflow.com/a/776523
+# - The GCC compiler recognizes this code as rotation and inserts a single
+#   x86/x86_64 rol/ror instruction since version 4.9.0 for `SomeUnsignedInt`.
+# - The CLANG compiler recognizes this code as rotation and inserts a single
+#   x86/x86_64 rol/ror instruction on version 8.0.0 and 11.0.0 to 12.0.0 for
+#   `SomeUnsignedInt`. CLANG from 7.0.0, 9.0.0 to 10.0.0 does not recognize for
+#   uint8 and uint16.
+func rotl[T: SomeUnsignedInt](value: T, rot: int32): T {.inline.} =
+  const mask: int32 = 8 * sizeof(value) - 1
+  (value shl rot) or (value shr (-rot and mask))
+
+func rotr[T: SomeUnsignedInt](value: T, rot: int32): T {.inline.} =
+  const mask: int32 = 8 * sizeof(value) - 1
+  (value shr rot) or (value shl (-rot and mask))
+
+func rotateLeftBits*(value: uint8, shift: range[0..8]): uint8 {.inline.} =
   ## Left-rotate bits in a 8-bits value.
   runnableExamples:
     doAssert rotateLeftBits(0b0000_0001'u8, 1) == 0b0000_0010'u8
     doAssert rotateLeftBits(0b0000_0001'u8, 2) == 0b0000_0100'u8
     doAssert rotateLeftBits(0b0100_0001'u8, 1) == 0b1000_0010'u8
     doAssert rotateLeftBits(0b0100_0001'u8, 2) == 0b0000_0101'u8
+  
+  when nimvm:
+    rotl(value, shift.int32)
+  else:
+    when useBuiltinsRotate:
+      when defined(gcc) or defined(icl):
+        builtin_rotl8(value.cuchar, shift.cint).uint8
+      elif defined(clang):
+        builtin_rotl8(value.cuchar, shift.cuchar).uint8
+      elif defined(vcc):
+        builtin_rotl8(value.cuchar, shift.cuchar).uint8
+    else:
+      rotl(value, shift.int32)
 
-  # using this form instead of the one below should handle any value
-  # out of range as well as negative values.
-  # result = (value shl amount) or (value shr (8 - amount))
-  # taken from: https://en.wikipedia.org/wiki/Circular_shift#Implementing_circular_shifts
-  let amount = amount and 7
-  result = (value shl amount) or (value shr ( (-amount) and 7))
-
-proc rotateLeftBits*(value: uint16;
-           amount: range[0..16]): uint16 {.inline, noSideEffect.} =
+func rotateLeftBits*(value: uint16, shift: range[0..16]): uint16 {.inline.} =
   ## Left-rotate bits in a 16-bits value.
   ##
   ## See also:
```

Comment:
IMO remove these "see also" clauses in the doc comments (these procs will be adjacent to eachother when documentation is rendered) and add runnable examples

Reply:
The runnableExamples for uint64 is pretty ugly. But it is done.